service/datasync/exec_data_sync_internal_test.go

```diff
@@ -0,0 +1,140 @@
+package datasync
+
+import (
+	"github.com/onflow/flow-go/engine/common/rpc/convert"
+	"github.com/onflow/flow-go/model/flow"
+	"testing"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/onflow/flow-archive/models/archive"
+	"github.com/onflow/flow-archive/testing/mocks"
+	"github.com/onflow/flow/protobuf/go/flow/entities"
+	execData "github.com/onflow/flow/protobuf/go/flow/executiondata"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewExecDataSync(t *testing.T) {
+	log := zerolog.Nop()
+	limit := uint(42)
+	blockIDs := mocks.GenericBlockIDs(4)
+
+	bed := entities.BlockExecutionData{
+		BlockId:            nil,
+		ChunkExecutionData: nil,
+	}
+	mockResponse := execData.GetExecutionDataByBlockIDResponse{
+		BlockExecutionData: &bed,
+	}
+
+	var mockApiClient execData.ExecutionDataAPIClient = mocks.MockExecAPIClient{
+		Response: &mockResponse,
+	}
+
+	streamer := NewExecDataSync(
+		log,
+		mockApiClient,
+		flow.Testnet.Chain(),
+		WithBufferSize(limit),
+		WithCatchupBlocks(blockIDs),
+	)
+
+	require.NotNil(t, streamer)
+	assert.NotZero(t, streamer.log)
+	assert.Equal(t, limit, streamer.limit)
+	assert.NotNil(t, streamer.blocks)
+	assert.NotNil(t, streamer.records)
+
+	for streamer.blocks.Len() > 0 {
+		assert.Contains(t, blockIDs, streamer.blocks.PopFront())
+	}
+}
+
+func TestStreamer_OnBlockFinalized(t *testing.T) {
+	block := mocks.GenericBlock
+	queue := archive.NewDeque()
+
+	streamer := &ExecDataSync{
+		log:    zerolog.Nop(),
+		blocks: queue,
+	}
+
+	streamer.OnBlockFinalized(block)
+
+	require.Equal(t, 1, queue.Len())
+	assert.Equal(t, queue.PopFront(), block.BlockID)
+}
+
+func TestStreamer_Next(t *testing.T) {
```

Comment:
can you also add a functional test that verifies the interaction between `poll` and `Next()`

e.g.
* instantiate the syncer and start it
* push several blocksIDs into the `blocks` and a single ExecData into `records`
* add a mock exec api client that returns a couple records then returns `NotFound`
* Call `Next()` until no more data is available and verify all of the expected blocks were retrieved